src/Storages/MergeTree/IntersectionsIndexes.h

```diff
@@ -0,0 +1,276 @@
+    PartBlockRange range;
+    String name;
+
+    bool operator==(const PartToRead & rhs) const
+    {
+        return range == rhs.range && name == rhs.name;
+    }
+
+    struct FindByBegin : std::binary_function<bool, Int64, PartToRead>
+    {
+        bool operator()(Int64 value, const PartToRead & rhs) const
+        {
+            return value < rhs.range.begin;
+        }
+    };
+
+    struct FindByEndGreater : std::binary_function<bool, Int64, PartToRead>
+    {
+        bool operator()(Int64 value, const PartToRead & lhs) const
+        {
+            return value > lhs.range.end;
+        }
+    };
+
+    struct CompareByBegin : std::binary_function<bool, PartToRead, PartToRead>
+    {
+        bool operator()(const PartToRead & lhs, const PartToRead & rhs) const
+        {
+            return lhs.range.begin < rhs.range.begin;
+        }
+    };
+
+
+    struct CompareByEnd : std::binary_function<bool, PartToRead, PartToRead>
+    {
+        bool operator()(const PartToRead & lhs, const PartToRead & rhs) const
+        {
+            return lhs.range.end < rhs.range.end;
+        }
+    };
+};
+
+
+class PartRangesIntersectionsIndex
+{
+public:
+    void addPart(PartToRead part)
+    {
+        by_begin.insert(part);
+        by_end.insert(part);
+    }
+
+    size_t numberOfIntersectionsWith(PartBlockRange range)
+    {
+        /// Find the first one that starts with a larger coordinate
+        auto right_iter = std::upper_bound(by_begin.begin(), by_begin.end(), range.end, PartToRead::FindByBegin());
+
+        /// Find the last that ends with a lower coordinate
+        auto left_iter = std::upper_bound(by_end.rbegin(), by_end.rend(), range.begin, PartToRead::FindByEndGreater());
+
+        size_t result = by_begin.size();
+        result -= std::distance(left_iter, by_end.rend());
+        result -= std::distance(right_iter, by_begin.end());
+
+        return result;
+    }
+
+    bool checkPartIsSuitable(PartToRead part)
+    {
+        auto number = numberOfIntersectionsWith(part.range);
+        if (number == 0)
+            return true;
+        if (number == 1)
+        {
+            auto it = by_begin.find(part);
+            if (it != by_begin.end() && it->name == part.name)
+                return true;
+        }
+
+        return false;
+    }
+
+    String describe() const
+    {
+        String result = "[";
+        for (const auto & it : by_begin)
+            result += fmt::format("({} {} {}), ", it.name, it.range.begin, it.range.end);
+        result += "]";
+        return result;
+    }
+
+
+    void checkConsistencyOrThrow() const
+    {
+        Int64 prev_end = std::numeric_limits<Int64>::min();
+        for (const auto & it : by_begin)
+        {
+            if (it.range.begin < prev_end)
+                throw Exception(ErrorCodes::LOGICAL_ERROR, "Intersecting marks");
+            prev_end = it.range.end;
+        }
+    }
+
+private:
+    std::multiset<PartToRead, PartToRead::CompareByBegin> by_begin;
```

Comment:
maybe std::set is enough

Reply:
Also discussed in chat:
1. Use one std::set : left_point -> PartToRead.
2. Use probably linear-time algorithm to check intersections is ok. Or, probably, we need to check if there is intersection at all: binary search left point and iterate forward, binary search right point and iterate backwards